feat: add `array/base/map`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `array/base/map` function including tests, benchmarks and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/21

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add array/base/map

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
Will open for final review, after a self-review some time today.

headlessNode commented on lib/node_modules/@stdlib/array/base/map/README.md:
````diff
@@ -0,0 +1,189 @@
+    x.map( fcn, thisArg )
+    ```
+
+-   The functions support array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).
````
```suggestion
-   The function supports array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/array/base/map/README.md:
```diff
@@ -0,0 +1,189 @@
+-   **index**: current element index.
+-   **arr**: input array.
+
+#### map.assign( x, y, fcn\[, thisArg] )
```
@headlessNode You need to support a stride and offset parameter. See, e.g., https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/cuevery-by#cueverybyassign-x-out-stride-offset-predicate-thisarg--.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/repl.txt:
```diff
@@ -0,0 +1,73 @@
+{{alias}}( x, fcn[, thisArg] )
```
Missing blank line.
```suggestion

{{alias}}( x, fcn[, thisArg] )
```

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,214 @@
+	* var y = map( x, scale );
+	* // returns {...}
+	*/
+	<T = unknown, U = unknown, V = unknown>( x: AccessorArrayLike<T>, fcn: Callback<T, U, V>, thisArg?: ThisParameterType<Callback<T, U, V>> ): AccessorArrayLike<U>;
```
@headlessNode You need to switch the order of the overloads, as an `AccessorArrayLike` object is also a collection.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,214 @@
+	* var x = toAccessorArray( ones( 4 ) );
+	* var y = toAccessorArray( zeros( x.length ) );
+	*
+	* var out = map.assign( x, y, scale );
+	* // returns {...}
```
```suggestion
	* var out = map.assign( toAccessorArray( x ), toAccessorArray( y ), scale );
	* // y => [ 10, 10, 10, 10 ]
```

Show the expected results.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,214 @@
+* var y = toAccessorArray( zeros( x.length ) );
+*
+* var out = map.assign( x, y, scale );
+* // returns {...}
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,153 @@
+* var x = arraylike2object( toAccessorArray( ones( 4 ) ) );
+* var y = arraylike2object( toAccessorArray( zeros( x.data.length ) ) );
+* var out = accessors( x, y, scale );
+* // returns {...}
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/main.js:
```diff
@@ -0,0 +1,161 @@
+*
+* var x = arraylike2object( toAccessorArray( ones( 4 ) ) );
+* var y = accessors( x, scale );
+* // returns {...}
```
Same comment.

kgryte replied:
This sort of example does not allow our doctesting framework to serve as lightweight unit tests. Here, the returns annotation is not particularly informative.

kgryte commented on lib/node_modules/@stdlib/array/base/map/test/test.main.js:
```diff
@@ -0,0 +1,136 @@
+	var x;
+
+	x = toAccessorArray( [ 1, 2, 3, 4 ] );
+	expected = arraylike2object( toAccessorArray( [ 10, 20, 30, 40 ] ) );
```
This isn't correct.

kgryte commented on lib/node_modules/@stdlib/array/base/map/test/test.main.js:
```diff
@@ -0,0 +1,136 @@
+
+	actual = map( x, scale );
+
+	t.deepEqual( actual, expected, 'returns expected value' );
```
I am surprised this passes. What is returned should not be the same as `arraylike2object`.

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/main.js:
```diff
@@ -0,0 +1,161 @@
+	}
+	obj = arraylike2object( x );
+	if ( obj.accessorProtocol ) {
+		return accessors( obj, fcn, thisArg );
```
This is incorrect. We should NOT be returning `obj`.

For both here and L155, you should have already preallocated a generic array at L151, and then you should pass that output array to `accessors` and `internal`, rather than allocating within those functions.

kgryte reviewed (changes requested):
Left an initial round of comments. This package needs a fair amount of refactoring before we can merge.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/array/base/map/README.md:
```diff
@@ -0,0 +1,191 @@
+-   **index**: current element index.
+-   **arr**: input array.
+
+#### map.assign( x, y, strideX, offsetX, fcn\[, thisArg] )
```
@headlessNode The stride and offset should apply to `y`, not `x`.

kgryte replied:
`assign` APIs conventionally have an output collection and an output stride and offset. The input array is assumed to be contiguous.

kgryte commented on lib/node_modules/@stdlib/array/base/map/benchmark/benchmark.assign.js:
```diff
@@ -0,0 +1,106 @@
+
+		b.tic();
+		for ( i = 0; i < b.iterations; i++ ) {
+			out = map( x, y, identity );
```
For the `assign` API, this is not the correct invocation.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/repl.txt:
```diff
@@ -0,0 +1,80 @@
+    [ 1, 2, 3, 4 ]
+
+
+{{alias}}.assign( x, y, strideX, offsetX, fcn[, thisArg] )
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/repl.txt:
```diff
@@ -0,0 +1,80 @@
+    y: Array|TypedArray|Object
+        Output array.
+
+    strideX: integer
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,208 @@
+	* var out = map.assign( toAccessorArray( x ), toAccessorArray( y ), scale );
+	* // y => [ 10, 10, 10, 10 ]
+	*/
+	assign<T = unknown, U = unknown, V = unknown>( x: AccessorArrayLike<T>, y: AccessorArrayLike<U>, fcn: Callback<T, U, V>, thisArg?: ThisParameterType<Callback<T, U, V>> ): AccessorArrayLike<U>;
```
This signature needs updating. Same below.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,208 @@
+	* Applies a callback function to elements in an input array and assigns results to elements in an output array.
+	*
+	* @param x - input array object
+	* @param y - output array object
```
Docs need updating.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/index.d.ts:
```diff
@@ -0,0 +1,208 @@
+	* var x = ones( 4 );
+	* var y = zeros( x.length );
+	*
+	* var out = map.assign( toAccessorArray( x ), toAccessorArray( y ), scale );
```
Needs updating.

kgryte commented on lib/node_modules/@stdlib/array/base/map/docs/types/test.ts:
```diff
@@ -0,0 +1,180 @@
+	const x = [ 1, 2, 3 ];
+	const y = [ 0, 0, 0 ];
+
+	map.assign( x, y, fcn ); // $ExpectType Collection<number>
```
All `assign` tests need updating.

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+	var ix;
+	var i;
+
+	ix = offsetX;
```
Implementation is not correct.

kgryte replied:
I already referenced this package, but you should take a closer look at its implementation: https://github.com/stdlib-js/stdlib/blob/0f30934d486c023154b676c08392e5d0134d130f/lib/node_modules/%40stdlib/array/base/cuevery-by/lib/assign.js#L57.

headlessNode commented:
ooff! my bad @kgryte Sorry for the overhead.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+* @private
+* @param {Collection} x - input array
+* @param {Collection} y - output array
+* @param {integer} strideX - stride length for x
```
The parameter names and descriptions still need updating.

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+* @private
+* @param {Object} x - input array object
+* @param {Object} y - output array object
+* @param {integer} strideX - stride length for x
```
Parameter names and descriptions still need updating.

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+	var i;
+
+	io = offsetX;
+	for ( i = 0; i < y.length; i++ ) {
```
You should not be basing the loop on `y`! Given that we support a stride and offset for `y`, the `length` of `y` could be much greater than `x`. This needs to be fixed here and below.

This should have been caught via unit tests, which suggests there is a gap in testing.

headlessNode commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+* @param {Collection} x - input array
+* @param {Collection} y - output array
+* @param {integer} strideX - stride length for x
+* @param {NonNegativeInteger} offsetX - starting index for x
```
```suggestion
* @param {integer} strideX - stride length for output array
* @param {NonNegativeInteger} offsetX - starting index for output array
```

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+*
+* @param {Collection} x - input array
+* @param {Collection} y - output array
+* @param {integer} strideX - stride length for x
```
Parameter names and descriptions.

headlessNode commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+* @param {Object} x - input array object
+* @param {Object} y - output array object
+* @param {integer} strideX - stride length for x
+* @param {NonNegativeInteger} offsetX - starting index for x
```
```suggestion
* @param {integer} strideX - stride length for output array
* @param {NonNegativeInteger} offsetX - starting index for output array
```

headlessNode commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+* @param {Collection} x - input array
+* @param {Collection} y - output array
+* @param {integer} strideX - stride length for x
+* @param {NonNegativeInteger} offsetX - starting index for x
```
```suggestion
* @param {integer} strideX - stride length for output array
* @param {NonNegativeInteger} offsetX - starting index for output array
```

kgryte commented on lib/node_modules/@stdlib/array/base/map/test/test.assign.js:
You need to add tests for the affects of different stride and offset values for each input/output data type. This is why the bug in the implementation was not caught. Similar to the tests in `blas/ext`. Strides, negative strides, offsets, etc.

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+	var i;
+
+	io = offsetX;
+	for ( i = 0; i < y.length; i++ ) {
```
```suggestion
	for ( i = 0; i < x.length; i++ ) {
```

headlessNode commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+	yset = y.accessors[ 1 ];
+
+	io = offsetX;
+	for ( i = 0; i < ydata.length; i++ ) {
```
```suggestion
	for ( i = 0; i < xdata.length; i++ ) {
```

headlessNode committed:
refactor: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
Merge branch 'array-base-map' of https://github.com/headlessNode/stdlib into array-base-map

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

headlessNode commented:
This was THE PR of mine with most amount of suggestions, ever. 😅

kgryte commented on lib/node_modules/@stdlib/array/base/map/lib/assign.js:
```diff
@@ -0,0 +1,164 @@
+* @private
+* @param {Collection} x - input array
+* @param {Collection} y - output array
+* @param {integer} strideX - stride length for output array
```
The parameter names are still incorrect. `strideX` is not what we want here, as the stride does not apply to `x`, but to `y`. Either rename to `stride` or `strideY`. The same applies for `offsetX` and throughout this file.

kgryte commented on lib/node_modules/@stdlib/array/base/map/test/test.assign.js:
```diff
@@ -0,0 +1,213 @@
+	}
+});
+
+tape( 'the function supports a `stride` parameter', function test( t ) {
```
We need to replicate these tests for each input argument type: "generic", "accessors", "typed array", and "array-like object". While internally they may share the same implementation, from a user POV that is not guaranteed. As such, unit tests should be comprehensive in terms of expected observable behavior. In general, we should not be writing tests based on insider knowledge.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte committed:
Merge branch 'develop' of https://github.com/stdlib-js/stdlib into pr/headlessNode/5346

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after clean-up.
